File: solutions/Leetcode_49/leetcode_49.py
from collections import defaultdict
from typing import List

class Solution1:
    def group_anagrams(self, strs):
        anagrams_dict = defaultdict(list)
        for word in strs:
            anagrams_dict[''.join(sorted(word))].append(word)
        return list(anagrams_dict.values())

class SolutionMINE:
    def group_anagrams(self, strs: List[str]) -> List[List[str]]:
        def hash(word):
            key = [0] * 26
            for x in word:
                key[ord(x) - 97] += 1
            return '.'.join(map(str, key))
        
        anagrams_dict = defaultdict(list)
        for word in strs:
            anagrams_dict[hash(word)].append(word)
        return list(anagrams_dict.values())

File: solutions/Leetcode_49/leetcode_49.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

class Solution1
{
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs)
    {
        std::unordered_map<std::string, std::vector<std::string>> anagramsMap;
        for (const std::string& word : strs)
        {
            std::string copyWord = word;
            std::sort(copyWord.begin(), copyWord.end());
            anagramsMap[copyWord].push_back(word);
        }
        std::vector<std::vector<std::string>> anagrams;
        for (const auto& keyAnagramPair : anagramsMap)
            anagrams.push_back(keyAnagramPair.second);
        return anagrams;
    }
};